net: pkt: Fix removal of empty buffers in net_pkt_pull()

jukkar opened the pull request:
Missed one corner case in my earlier PR.

If we have removed first net_buf, then we must not restore the
original cursor as that will point to wrong head net_buf.
Add also unit test to check that the packets are removed
properly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: pkt: Fix removal of empty buffers in net_pkt_pull()

If we have removed first net_buf, then we must not restore the
original cursor as that will point to wrong head net_buf.
Add also unit test to check that the packets are removed
properly.
Clarify the documentation that we are removing data from
beginning of the function, also document that the cursor
is reset after this call.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1873,6 +1873,7 @@ int net_pkt_pull(struct net_pkt *pkt, size_t length)
+	bool restore = true;
 
 	if (!pkt->buffer) {
 		return -ENOBUFS;
```
btw you added that check in the previous patch, but that's superfluous since if the packet has no buffer the cursor should not point to anything (if it does, it's a bug induced by the caller of this function.), it should be NULL. Thus why there is a test at line 1889.

jukkar commented on subsys/net/ip/net_pkt.c:
```diff
@@ -1910,15 +1911,18 @@ int net_pkt_pull(struct net_pkt *pkt, size_t length)
-				net_pkt_cursor_init(pkt);
 			}
+
+			net_pkt_cursor_init(pkt);
```
The `net_pkt_pull()` only pulls data from the beginning of the pkt, actually I wonder does it makes sense to restore the cursor at all.

jukkar replied:
Clarification, the documentation says that the data is pulled from cursor, I probably missed something but did not quite understand how it could work by looking the original code.

tbursztyka replied:
well again, forgot how this api works ... 
the doc say "Remove data from the packet at current location" so it removes what's in front of the cursor (I thought the opposite for some reason).

then no your patch is right, if the cursor was just initialized at the beginning of the pkt, but user request to remove <size of a net_buf>, then restoring is wrong indeed. 

So maybe it should not backup/restore, and always initialize at the end.

jukkar replied:
> So maybe it should not backup/restore, and always initialize at the end.

Yes, it looks like the restoration can cause issues. I will experiment a bit about this.

jukkar commented:
I clarified the documentation and removed the cursor restoration as the cursor might point to wrong place after the pull so it would give false sense of security to rely on its value.

tbursztyka commented on include/net/net_pkt.h:
```diff
@@ -1801,10 +1801,9 @@ size_t net_pkt_remaining_data(struct net_pkt *pkt);
 
 /**
- * @brief Remove data from the packet at current location
+ * @brief Remove data from the beginning of the packet
```
no that's not changing. And actually we don't want to, see ipv6_fragment.c.

tbursztyka commented on include/net/net_pkt.h:
```diff
@@ -1801,10 +1801,9 @@ size_t net_pkt_remaining_data(struct net_pkt *pkt);
  *
- * @details net_pkt's cursor should be properly initialized and,
- *          eventually, properly positioned using net_pkt_skip/read/write.
+ * @details net_pkt's cursor is reset by this function
```
Yes, and that will let you optimize where net_pkt_pull() is called, like in ipv6_fragment.c : you'll be able to remove the net_pkt_cursor_init() which is being called right after net_pkt_pull() there.

jukkar committed:
net: ipv6_fragment: Remove extra net_pkt_cursor_init() call

The net_pkt_pull() will reset the cursor after its call so no
need to call it here.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Unit tests pass but I am seeing weird issues with real program, investigating.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -1803,8 +1803,7 @@ int net_pkt_update_length(struct net_pkt *pkt, size_t length);
  *
- * @details net_pkt's cursor should be properly initialized and,
- *          eventually, properly positioned using net_pkt_skip/read/write.
+ * @details net_pkt's cursor is reset by this function
```
Yep, restored the text.

jukkar commented:
Works now. The issue was in my code that was calling the net_pkt_pull().
